### PR TITLE
Update web backend get connection to properly handle field selection and schema refresh

### PR DESCRIPTION
**Pull request by mfsiega-airbyte**
## What
Make the web backend handler properly handle field selection when there is a schema refresh.

## How
Only include a selected field if:

`is in the newly discovered schema` AND `it was either originally selected OR not in the originally discovered schema at all`.

TODO before merging: add another unit test for the case where a field is removed in the newly-discovered schema.

## Recommended reading order
Reading the new unit test might make the intended logic clearer, and therefore easier to review the actual code change.

### CODE REVIEW

File: airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java [hunk 1]
```diff
@@ -357,7 +360,7 @@ public WebBackendConnectionRead webBackendGetConnection(final WebBackendConnecti
       connection.setStatus(refreshedCatalog.get().getConnectionStatus());
     } else if (catalogUsedToMakeConfiguredCatalog.isPresent()) {
       // reconstructs a full picture of the full schema at the time the catalog was configured.
-      syncCatalog = updateSchemaWithDiscovery(configuredCatalog, catalogUsedToMakeConfiguredCatalog.get());
+      syncCatalog = updateSchemaWithDiscovery(configuredCatalog, catalogUsedToMakeConfiguredCatalog.get(), catalogUsedToMakeConfiguredCatalog.get());
```
Comment (pmossman) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java:363]: Why are we just passing the same `catalogUsedToMakeConfiguredCatalog.get()` in twice?

Comment (pmossman) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java:363]: Oh hm, I think I see why but it's confusing, maybe in the case where there isn't a refreshed catalog, we should call a different, simpler method that takes advantage of the fact that the original discovered catalog is equal to the discovered catalog?

Comment (mfsiega-airbyte) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java]: Yeah I can see how this is confusing.

It wouldn't actually simplify much to have two different methods - essentially it's _mostly_ saying "reconstruct it by looking at what we persisted versus the latest catalog". This PR adds a third element: (1) what we persisted; (2) the whole catalog when we persisted it; (3) the latest catalog. It's just that in the case where the catalog hasn't been rediscovered, (2) and (3) are actually the same. But otherwise the logic stays the same.

I do agree that it stands out as _weird_ to pass the same thing twice like that, so I added a helper method in the hopes of better readability.

WDYT?

Comment (pmossman) [reply airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java]: Nice! Makes sense, the helper makes it much clearer

File: airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java [hunk 2]
```diff
@@ -433,16 +452,29 @@ protected static AirbyteCatalog updateSchemaWithDiscovery(final AirbyteCatalog o
...
+          // If field selection is enabled, populate the selected fields.
+          final Set<String> originallyDiscovered = new HashSet<>();
+          final Set<String> refreshDiscovered = new HashSet<>();
+          // NOTE: by only taking the first element of the path, we're restricting to top-level fields.
+          final Set<String> originallySelected = new HashSet<>(
+              originalConfiguredStream.getConfig().getSelectedFields().stream().map((field) -> field.getFieldPath().get(0)).toList());
+          originalDiscoveredStream.getStream().getJsonSchema().findPath("properties").fieldNames()
+              .forEachRemaining((name) -> originallyDiscovered.add(name));
+          stream.getJsonSchema().findPath("properties").fieldNames().forEachRemaining((name) -> refreshDiscovered.add(name));
+          // We include a selected field if it:
```
Comment (pmossman) [inline airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java:469]: great comments, would be hard to follow without them!

File: airbyte-server/src/test/java/io/airbyte/server/handlers/WebBackendConnectionsHandlerTest.java [hunk 3]
```diff
@@ -439,12 +447,109 @@ void testWebBackendGetConnectionWithDiscoveryAndNewSchemaBreakingChange() throws
...
+                    .addField(new FieldAdd().schema(newFieldSchema)))))
+            .catalog(newCatalogToDiscover)
+            .breakingChange(false)
+            .connectionStatus(ConnectionStatus.ACTIVE);
+    when(schedulerHandler.discoverSchemaForSourceFromSourceId(any())).thenReturn(schemaRead);
+
+    final WebBackendConnectionRead result = testWebBackendGetConnection(true, connectionRead,
+        operationReadList);
+
+    // We expect the discovered catalog with two fields selected: the one that was originally selected,
```
Comment (pmossman) [inline airbyte-server/src/test/java/io/airbyte/server/handlers/WebBackendConnectionsHandlerTest.java:498]: Also really appreciate the comments describing how the test works!

**Review by pmossman — APPROVED** [review]
Nice! Really helpful comments that help navigate the complexity. Also test cases look solid, great stuff!

**Comment by marcelopio** [comment]
This change broke the "Refresh Schema" button for a lot of connections on my install.

```
2022-12-23 13:47:04 [1;31mERROR[m i.a.s.a.ApiHelper(execute):28 - Unexpected Exception
java.util.NoSuchElementException: No value present
	at java.util.Optional.get(Optional.java:143) ~[?:?]
	at io.airbyte.server.handlers.WebBackendConnectionsHandler.webBackendGetConnection(WebBackendConnectionsHandler.java:352) ~[io.airbyte-airbyte-server-0.40.26.jar:?]
	at io.airbyte.server.apis.WebBackendApiController.lambda$webBackendGetConnection$2(WebBackendApiController.java:51) ~[io.airbyte-airbyte-server-0.40.26.jar:?]
	at io.airbyte.server.apis.ApiHelper.execute(ApiHelper.java:18) ~[io.airbyte-airbyte-server-0.40.26.jar:?]
	at io.airbyte.server.apis.WebBackendApiController.webBackendGetConnection(WebBackendApiController.java:51) ~[io.airbyte-airbyte-server-0.40.26.jar:?]
	at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:104) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:578) ~[?:?]
	at org.glassfish.jersey.server.model.internal.ResourceMethodInvocationHandlerFactory.lambda$static$0(ResourceMethodInvocationHandlerFactory.java:52) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher$1.run(AbstractJavaResourceMethodDispatcher.java:124) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.invoke(AbstractJavaResourceMethodDispatcher.java:167) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.internal.JavaResourceMethodDispatcherProvider$TypeOutInvoker.doDispatch(JavaResourceMethodDispatcherProvider.java:219) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.internal.AbstractJavaResourceMethodDispatcher.dispatch(AbstractJavaResourceMethodDispatcher.java:79) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.invoke(ResourceMethodInvoker.java:469) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:391) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.model.ResourceMethodInvoker.apply(ResourceMethodInvoker.java:80) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.ServerRuntime$1.run(ServerRuntime.java:253) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:248) ~[jersey-common-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors$1.call(Errors.java:244) ~[jersey-common-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:292) ~[jersey-common-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:274) ~[jersey-common-2.31.jar:?]
	at org.glassfish.jersey.internal.Errors.process(Errors.java:244) ~[jersey-common-2.31.jar:?]
	at org.glassfish.jersey.process.internal.RequestScope.runInScope(RequestScope.java:265) ~[jersey-common-2.31.jar:?]
	at org.glassfish.jersey.server.ServerRuntime.process(ServerRuntime.java:232) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.server.ApplicationHandler.handle(ApplicationHandler.java:680) ~[jersey-server-2.31.jar:?]
	at org.glassfish.jersey.servlet.WebComponent.serviceImpl(WebComponent.java:394) ~[jersey-container-servlet-core-2.31.jar:?]
	at org.glassfish.jersey.servlet.WebComponent.service(WebComponent.java:346) ~[jersey-container-servlet-core-2.31.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:366) ~[jersey-container-servlet-core-2.31.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:319) ~[jersey-container-servlet-core-2.31.jar:?]
	at org.glassfish.jersey.servlet.ServletContainer.service(ServletContainer.java:205) ~[jersey-container-servlet-core-2.31.jar:?]
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:763) ~[jetty-servlet-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:569) ~[jetty-servlet-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1377) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:507) ~[jetty-servlet-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1292) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.server.Server.handle(Server.java:501) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:383) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:556) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:375) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:273) ~[jetty-server-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) ~[jetty-io-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) ~[jetty-io-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104) ~[jetty-io-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336) ~[jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313) ~[jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171) ~[jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129) ~[jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375) ~[jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806) ~[jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938) ~[jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at java.lang.Thread.run(Thread.java:1589) ~[?:?]
```

This is because of this comment, some lines below the change:
https://github.com/airbytehq/airbyte/blob/825278300b4e580fc3e54c63df6fadce4fc2dba1/airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java#L351-L358

There should be a check to see if `catalogUsedToMakeConfiguredCatalog` is also present here:
https://github.com/airbytehq/airbyte/blob/825278300b4e580fc3e54c63df6fadce4fc2dba1/airbyte-server/src/main/java/io/airbyte/server/handlers/WebBackendConnectionsHandler.java#L343
